### PR TITLE
[CALCITE-4897] Implicit type conversion is not complete for set operation in DML

### PR DESCRIPTION
**Pull request by JiajunBernoulli**
# insert union case

 Here is the sql:
```
insert into t3 
select 'a', 1.0, 1 union 
select 'b', 2, 2 union 
select 'c', 3.0, CAST(3 AS SMALLINT) union 
select 'd', 4.0, 4 union 
select 'e', 5.0, 5 
```
but the validated sql is :
```
INSERT INTO `CATALOG`.`SALES`.`T3`
SELECT 'a', CAST(1.0 AS INTEGER), CAST(1 AS SMALLINT) UNION 
SELECT 'b', 2, CAST(2 AS SMALLINT) UNION -- Encountered an integer and ended early
SELECT 'c', 3.0, CAST(3 AS SMALLINT) UNION -- Encountered an samlint and ended early
SELECT 'd', 4.0, 4 UNION -- should be cast  
SELECT 'e', 5.0, 5 -- should be cast 
```
This is why CALCITE-4458 changed '&&' to '||', the right validated sql is:
```
INSERT INTO `CATALOG`.`SALES`.`T3`
SELECT 'a', CAST(1.0 AS INTEGER), CAST(1 AS SMALLINT)
UNION
SELECT 'b', 2, CAST(2 AS SMALLINT)
UNION
SELECT 'c', CAST(3.0 AS INTEGER), CAST(3 AS SMALLINT)
UNION
SELECT 'd', CAST(4.0 AS INTEGER), CAST(4 AS SMALLINT)
UNION
SELECT 'e', CAST(5.0 AS INTEGER), CAST(5 AS SMALLINT)
 ```

# insert values case
Here is the sql
```
insert into t3 values ('a', 1.0, 1), ('b', 2, 2), ('c', 3.0, CAST(3 AS SMALLINT)), ('d', 4.0, 4), ('e', 5.0, 5)
```
 the validated sql is :
```
INSERT INTO `CATALOG`.`SALES`.`T3`
VALUES ROW('a', CAST(1.0 AS INTEGER), CAST(1 AS SMALLINT)),
ROW('b', 2, CAST(2 AS SMALLINT)),-- Encountered an integer and ended early
ROW('c', 3.0, CAST(3 AS SMALLINT)),Encountered an samllint and ended early
ROW('d', 4.0, 4),
ROW('e', 5.0, 5) 
```
the right validated sql is:
```
INSERT INTO `CATALOG`.`SALES`.`T3`
VALUES ROW('a', CAST(1.0 AS INTEGER), CAST(1 AS SMALLINT)),
ROW('b', 2, CAST(2 AS SMALLINT)),
ROW('c', CAST(3.0 AS INTEGER), CAST(3 AS SMALLINT)),
ROW('d', CAST(4.0 AS INTEGER), CAST(4 AS SMALLINT)),
ROW('e', CAST(5.0 AS INTEGER), CAST(5 AS SMALLINT))
```

### CODE REVIEW

**Comment by beyond1920** [comment]
@JiajunBernoulli LGTM. I only left a minor comment.

File: core/src/main/java/org/apache/calcite/sql/validate/implicit/TypeCoercionImpl.java [hunk 1]
```diff
@@ -104,14 +104,17 @@ public TypeCoercionImpl(RelDataTypeFactory typeFactory, SqlValidator validator)
...
-          return false;
+          coercedValues = false;
         }
       }
-      updateInferredColumnType(
-          requireNonNull(scope, "scope"), query, columnIndex, targetType);
-      return true;
+      if (coercedValues) {
+        updateInferredColumnType(
+            requireNonNull(scope, "scope"), query, columnIndex, targetType);
```
Comment (danny0405) [inline core/src/main/java/org/apache/calcite/sql/validate/implicit/TypeCoercionImpl.java:115]: What's the difference here ? The original code does the same behavior: `#updateInferredColumnType` only if all the operands are coerced.

File: core/src/main/java/org/apache/calcite/sql/validate/implicit/TypeCoercionImpl.java [hunk 2]
```diff
@@ -121,8 +124,8 @@ public TypeCoercionImpl(RelDataTypeFactory typeFactory, SqlValidator validator)
       // Set operations are binary for now.
       final SqlCall operand0 = ((SqlCall) query).operand(0);
       final SqlCall operand1 = ((SqlCall) query).operand(1);
-      final boolean coerced = rowTypeCoercion(scope, operand0, columnIndex, targetType)
-          && rowTypeCoercion(scope, operand1, columnIndex, targetType);
+      boolean coerced = rowTypeCoercion(scope, operand0, columnIndex, targetType);
```
Comment (danny0405) [inline core/src/main/java/org/apache/calcite/sql/validate/implicit/TypeCoercionImpl.java:127]: Change `&&` directly to `||` ?

File: core/src/test/java/org/apache/calcite/test/TypeCoercionTest.java [hunk 3]
```diff
@@ -894,6 +894,13 @@ public MockCatalogReader init() {
       t2.addColumn("t2_binary", binaryType);
       t2.addColumn("t2_boolean", booleanType);
       registerTable(t2);
+
+      final MockTable t3 =
+          MockTable.create(this, tSchema, "T3", false, 7.0, null);
+      t3.addColumn("t3_varchar20", varchar20Type);
+      t3.addColumn("t3_int", intType);
```
Comment (danny0405) [inline core/src/test/java/org/apache/calcite/test/TypeCoercionTest.java:901]: Why add a new table here ? 't1' and 't2' have all the data types you needed.

**Review by danny0405 — CHANGES_REQUESTED** [review]

File: core/src/test/java/org/apache/calcite/test/TypeCoercionTest.java (hunk 3 above)
Comment (JiajunBernoulli) [reply core/src/test/java/org/apache/calcite/test/TypeCoercionTest.java:901]: I'm worried that too **many columns** will affect everyone's reading, so I only use **a few columns.** If you think it is more appropriate to use the existing table, I can change it later.

File: core/src/main/java/org/apache/calcite/sql/validate/implicit/TypeCoercionImpl.java (hunk 1 above)
Comment (JiajunBernoulli) [reply core/src/main/java/org/apache/calcite/sql/validate/implicit/TypeCoercionImpl.java:115]: I thought so at first. But many test will be failed, for example:
```
 sql("select 1 from (values(true)) union values (select '1' from (values (true)) as tt)")
        .type("RecordType(VARCHAR EXPR$0) NOT NULL");
```
```
org.opentest4j.AssertionFailedError: expected: <RecordType(VARCHAR EXPR$0) NOT NULL> 
but was: <RecordType(VARCHAR NOT NULL EXPR$0) NOT NULL>
```
I think it is only executed when coerced values is true, like the following code
```      
// Update the nested SET operator node type.
if (coerced) {
  updateInferredColumnType(
      requireNonNull(scope, "scope"), query, columnIndex, targetType);
}
```

File: core/src/main/java/org/apache/calcite/sql/validate/implicit/TypeCoercionImpl.java (hunk 2 above)
Comment (JiajunBernoulli) [reply core/src/main/java/org/apache/calcite/sql/validate/implicit/TypeCoercionImpl.java:127]: This will not fix the problem. Because when `rowTypeCoercion(scope, operand0, columnIndex, targetType)=false`,  `rowTypeCoercion(scope, operand1, columnIndex, targetType) `will not be run.
I will use `|`, do you agree?
```
final boolean coerced = rowTypeCoercion(scope, operand0, columnIndex, targetType)
    | rowTypeCoercion(scope, operand1, columnIndex, targetType);
```

File: core/src/main/java/org/apache/calcite/sql/validate/implicit/TypeCoercionImpl.java (hunk 1 above)
Comment (danny0405) [reply core/src/main/java/org/apache/calcite/sql/validate/implicit/TypeCoercionImpl.java:115]: > org.opentest4j.AssertionFailedError: expected: <RecordType(VARCHAR EXPR$0) NOT NULL> 
but was: <RecordType(VARCHAR NOT NULL EXPR$0) NOT NULL>

Shouldn't this be the correct behavior ? Two constant literals are type nullable false, and implicit type coercion does not solve nullability in-consistency.

File: core/src/main/java/org/apache/calcite/sql/validate/implicit/TypeCoercionImpl.java (hunk 2 above)
Comment (danny0405) [reply core/src/main/java/org/apache/calcite/sql/validate/implicit/TypeCoercionImpl.java:127]: Yeah, you code is correct, we should not add any short-cut logic here.

File: core/src/test/java/org/apache/calcite/test/TypeCoercionConverterTest.java [hunk 4]
```diff
@@ -133,4 +162,24 @@
...
+    } catch (Exception e) {
+      throw TestUtil.rethrow(e);
+    }
+    final RelDataTypeFactory typeFactory = tester.getValidator().getTypeFactory();
+    final Prepare.CatalogReader catalogReader =
+        tester.createCatalogReader(typeFactory);
+    final SqlValidator validator =
+        tester.createValidator(catalogReader, typeFactory);
+
+    final String actual = validator.validate(sqlQuery).toString();
```
Comment (danny0405) [inline core/src/test/java/org/apache/calcite/test/TypeCoercionConverterTest.java:182]: Somehow we need some refactoring for this testing infrustructure, for e.g, a `Sql` util, we can do that in another PR.

File: core/src/test/java/org/apache/calcite/test/TypeCoercionTest.java (hunk 3 above)
Comment (danny0405) [reply core/src/test/java/org/apache/calcite/test/TypeCoercionTest.java:901]: Yeah, i'm sliding to reuse the original tables.

Comment (JiajunBernoulli) [reply core/src/test/java/org/apache/calcite/test/TypeCoercionTest.java]: I deleted `t3` and used `t1`.

File: core/src/main/java/org/apache/calcite/sql/validate/implicit/TypeCoercionImpl.java (hunk 2 above)
Comment (JiajunBernoulli) [reply core/src/main/java/org/apache/calcite/sql/validate/implicit/TypeCoercionImpl.java]: done

File: core/src/main/java/org/apache/calcite/sql/validate/implicit/TypeCoercionImpl.java (hunk 1 above)
Comment (JiajunBernoulli) [reply core/src/main/java/org/apache/calcite/sql/validate/implicit/TypeCoercionImpl.java]: I modified the expected values of failed unit tests. They should all be not null. Are there any other questions?

File: core/src/test/java/org/apache/calcite/test/TypeCoercionConverterTest.java (hunk 4 above)
Comment (JiajunBernoulli) [reply core/src/test/java/org/apache/calcite/test/TypeCoercionConverterTest.java:192]: I'm interested in it. `checkValidateSqlEquals` function can be abstracted. 
Do you think it's more appropriate to put it in the parent class `SqlToRelTestBase` or create an interface `SqlToValidateTestBase`?

File: core/src/test/java/org/apache/calcite/test/TypeCoercionConverterTest.java [hunk 5]
```diff
@@ -133,4 +172,24 @@
...
+
+  private void checkValidateSqlEquals(String sql) {
+    Objects.requireNonNull(sql, "sql");
+    final SqlNode sqlQuery;
+    try {
+      sqlQuery = tester.parseQuery(sql);
+    } catch (RuntimeException | Error e) {
+      throw e;
+    } catch (Exception e) {
+      throw TestUtil.rethrow(e);
```
Comment (vlsi) [inline core/src/test/java/org/apache/calcite/test/TypeCoercionConverterTest.java:184]: Consider using `catch(Throwable e) { TestUtil.rethrow(e, "sql=" + sql) }`
That would be shorter, it would catch all throwables, and it would add the SQL in question

File: core/src/test/java/org/apache/calcite/test/TypeCoercionConverterTest.java (hunk 4 above)
Comment (vlsi) [reply core/src/test/java/org/apache/calcite/test/TypeCoercionConverterTest.java:192]: @JiajunBernoulli , have you seen https://lists.apache.org/thread/3l9y4fv31kthmpn9pbzqdmfp80yrjphf by @julianhyde ?

File: core/src/test/java/org/apache/calcite/test/TypeCoercionConverterTest.java (hunk 5 above)
Comment (JiajunBernoulli) [reply core/src/test/java/org/apache/calcite/test/TypeCoercionConverterTest.java]: I have modified it, thank you very much for your advice.

File: core/src/test/java/org/apache/calcite/test/TypeCoercionConverterTest.java (hunk 4 above)
Comment (JiajunBernoulli) [reply core/src/test/java/org/apache/calcite/test/TypeCoercionConverterTest.java]: I didn't know before. I will continue to pay attention to it

File: core/src/main/java/org/apache/calcite/sql/validate/implicit/TypeCoercionImpl.java [hunk 6]
```diff
@@ -121,8 +122,9 @@ public TypeCoercionImpl(RelDataTypeFactory typeFactory, SqlValidator validator)
       // Set operations are binary for now.
       final SqlCall operand0 = ((SqlCall) query).operand(0);
       final SqlCall operand1 = ((SqlCall) query).operand(1);
-      final boolean coerced = rowTypeCoercion(scope, operand0, columnIndex, targetType)
-          && rowTypeCoercion(scope, operand1, columnIndex, targetType);
+      @SuppressWarnings("ShortCircuitBoolean")
+      boolean coerced = rowTypeCoercion(scope, operand0, columnIndex, targetType)
+          | rowTypeCoercion(scope, operand1, columnIndex, targetType);
       // Update the nested SET operator node type.
```
Comment (danny0405) [inline core/src/main/java/org/apache/calcite/sql/validate/implicit/TypeCoercionImpl.java:128]: We should not use `|` in the logical operator.

Comment (JiajunBernoulli) [reply core/src/main/java/org/apache/calcite/sql/validate/implicit/TypeCoercionImpl.java:127]: I didn't find a more suitable way, just modify the original code.

**Review by danny0405 — APPROVED** [review]
Overall looks good now, would refactor the test code when i merge it ~

**Comment by JiajunBernoulli** [comment]
> Overall looks good now, would refactor the test code when i merge it ~
Hello, the PR has been approved for five months. Do you have any plans to refactor the unit test recently? I want to create a new JIRA ticket to refactor the unit test this month. After refactoring, merge the commit.

**Comment by JiajunBernoulli** [comment]
@danny0405  Hello, I refactor the tests and rebase main branch. Ci failed for other reasons, so the PR is ready for Merge. Would you please merge it?  
BTW, I delete insert values case because many conflicts need to discuss, I will create new jiar ticket to solve it. The PR only solve problem for set operation.

**Comment by JiajunBernoulli** [comment]
@danny0405  Would you please review the pr again?

**Comment by JiajunBernoulli** [comment]
@rubenada, You left a [comment](https://github.com/apache/calcite/pull/2320#issuecomment-757456022) on CALCITE-4458(The solution is not correct and lacks unit tests).  
I believe you should be familiar with implicit type cast. The PR has been approved by dannny, but I can't contact him now. Would you please review this PR? Thanks.

**Comment by rubenada** [comment]
LGTM

**Review by danny0405 — APPROVED** [review]
+1， thanks for the fix, looks good from my side.

**Comment by JiajunBernoulli** [comment]
@danny0405, @rubenada . Thanks for your review.
